test/features/build_server/compiler_interface_test.dart

```diff
@@ -0,0 +1,319 @@
+import 'dart:async';
+import 'dart:io';
+import 'dart:isolate';
+
+import 'package:flyde/core/async/connect.dart';
+import 'package:flyde/core/async/interface.dart';
+import 'package:flyde/core/fs/compiler/installed_compiler.dart';
+import 'package:flyde/core/fs/configs/compiler_config.dart';
+import 'package:flyde/core/fs/file_extension.dart';
+import 'package:flyde/core/fs/search_directory.dart';
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:flyde/core/networking/protocol/compile_status.dart';
+import 'package:flyde/features/build_server/cache/project_cache.dart';
+import 'package:flyde/features/build_server/compiler_interface.dart';
+import 'package:path/path.dart' as p;
+import 'package:test/test.dart';
+
+import '../../helpers/clear_test_cache_directory.dart';
+import '../../helpers/create_dummy_project_cache.dart';
+import '../../helpers/value_hook.dart';
+
+Future<List<CompileStatusMessage>> _buildProject(
+  MainInterface interface,
+  List<SourceFile> files,
+  Map<String, String> fileMap,
+  CompilerConfig config,
+  ProjectCache cache, {
+  bool skipUpdateFiles = false,
+}) async {
+  final Completer<List<CompileStatusMessage>> completer = Completer();
+  final List<CompileStatusMessage> messageCache = [];
+
+  interface.onStateUpdate = (CompileStatusMessage m) {
+    messageCache.add(m);
+    if (m.status == CompileStatus.done) {
+      completer.complete(messageCache);
+    }
+    if (m.status == CompileStatus.failed) {
+      completer.completeError(StateError(m.payload));
+    }
+  };
+
+  await interface.init(fileMap, config, cache);
+
+  if (!skipUpdateFiles) {
+    final requiredFiles = await interface.sync(fileMap, config);
+
+    for (final file in requiredFiles) {
+      final srcFile = files.singleWhere((element) => element.id == file);
+      await interface.update(srcFile);
+    }
+  }
+
+  await interface.build();
+
+  return await completer.future;
+}
+
+Future<void> main() async {
+  MainInterface? interface;
```

Comment:
```suggestion
  late MainInterface interface;
```
This would avoid the need for `!`. We can pretty much guarantee that it wont be null